From 36315cbe2b3c9d1c1b7508d9494a251eddbc4452 Mon Sep 17 00:00:00 2001
From: John Ralls <jralls@ceridwen.us>
Date: Thu, 14 Oct 2021 12:41:05 -0700
Subject: [PATCH] [quartz]Use the absolute value of Scale values for
 CGContextScaleSCM

In macOS-12.sdk CGContextConverSizeToDeviceSpace returns a negative
height and passing that to CGContextScaleCTM in turn causes the cairo
surface to draw outside the window where it can't be seen. Passing the
absolute values of the scale factors fixes the display on macOS 12 without
affecting earlier macOS versions.
---
 gdk/quartz/gdkwindow-quartz.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/gdk/quartz/gdkwindow-quartz.c b/gdk/quartz/gdkwindow-quartz.c
index 1a3389bb32..a5c5c31945 100644
--- gdk/quartz/gdkwindow-quartz.c
+++ gdk/quartz/gdkwindow-quartz.c
@@ -183,8 +183,7 @@ gdk_window_impl_quartz_get_context (GdkWindowImplQuartz *window_impl,
    * in gdk_quartz_ref_cairo_surface () */
   scale = CGContextConvertSizeToDeviceSpace (cg_context,
                                              CGSizeMake (1.0, 1.0));
-  CGContextScaleCTM (cg_context, 1.0 / scale.width, 1.0 / scale.height);
-
+  CGContextScaleCTM (cg_context, 1.0 / fabs(scale.width), 1.0 / fabs(scale.height));
   return cg_context;
 }

--
GitLab

From a752e338381bc37dbe8d4c04ec23e4f6fd911b30 Mon Sep 17 00:00:00 2001
From: John Ralls <jralls@ceridwen.us>
Date: Wed, 8 Sep 2021 18:01:50 -0700
Subject: [PATCH] Update Quartz version detection for macOS 12.

---
 gdk/quartz/gdkglobals-quartz.c | 16 +++++++---------
 gdk/quartz/gdkquartz.h         |  3 ++-
 2 files changed, 9 insertions(+), 10 deletions(-)

diff --git a/gdk/quartz/gdkglobals-quartz.c b/gdk/quartz/gdkglobals-quartz.c
index 72aed1c92d..30649efe8a 100644
--- gdk/quartz/gdkglobals-quartz.c
+++ gdk/quartz/gdkglobals-quartz.c
@@ -29,28 +29,26 @@ GdkWindow *_gdk_root = NULL;
 GdkOSXVersion
 gdk_quartz_osx_version (void)
 {
-  static gint32 minor = GDK_OSX_UNSUPPORTED;
+  static gint32 vkey = GDK_OSX_UNSUPPORTED;

-  if (minor == GDK_OSX_UNSUPPORTED)
+  if (vkey == GDK_OSX_UNSUPPORTED)
     {
 #if MAC_OS_X_VERSION_MIN_REQUIRED < 101000
-      OSErr err = Gestalt (gestaltSystemVersionMinor, (SInt32*)&minor);
+      OSErr err = Gestalt (gestaltSystemVersionMinor, (SInt32*)&vkey);

       g_return_val_if_fail (err == noErr, GDK_OSX_UNSUPPORTED);
 #else
       NSOperatingSystemVersion version;

       version = [[NSProcessInfo processInfo] operatingSystemVersion];
-      minor = version.minorVersion;
-      if (version.majorVersion == 11)
-        minor += 16;
+      vkey = version.majorVersion == 10 ? version.minorVersion : version.majorVersion + 5;
 #endif
     }

-  if (minor < GDK_OSX_MIN)
+  if (vkey < GDK_OSX_MIN)
     return GDK_OSX_UNSUPPORTED;
-  else if (minor > GDK_OSX_CURRENT)
+  else if (vkey > GDK_OSX_CURRENT)
     return GDK_OSX_NEW;
   else
-    return minor;
+    return vkey;
 }
diff --git a/gdk/quartz/gdkquartz.h b/gdk/quartz/gdkquartz.h
index dd1b4f41af..2d53dcbce6 100644
--- gdk/quartz/gdkquartz.h
+++ gdk/quartz/gdkquartz.h
@@ -42,7 +42,8 @@ typedef enum
   GDK_OSX_MOJAVE = 14,
   GDK_OSX_CATALINA = 15,
   GDK_OSX_BIGSUR = 16,
-  GDK_OSX_CURRENT = 15,
+  GDK_OSX_MONTEREY = 17,
+  GDK_OSX_CURRENT = 17,
   GDK_OSX_NEW = 99
 } GdkOSXVersion;

--
GitLab

